Sources/HummingbirdWSCore/PerMessageDeflateExtension.swift

```diff
@@ -0,0 +1,278 @@
+            sendNoContextTakeover: serverNoContextTakeoverParam || self.serverNoContextTakeover,
+            compressionLevel: self.compressionLevel,
+            memoryLevel: self.memoryLevel
+        )
+    }
+}
+
+/// PerMessageDeflate websocket extension
+///
+/// Uses deflate to compress messages sent across a WebSocket
+/// See RFC 7692 for more details https://www.rfc-editor.org/rfc/rfc7692
+struct PerMessageDeflateExtension: HBWebSocketExtension {
+    enum SendState: Sendable {
+        case idle
+        case sendingMessage
+    }
+
+    struct Configuration: Sendable {
+        let receiveMaxWindow: Int?
+        let receiveNoContextTakeover: Bool
+        let sendMaxWindow: Int?
+        let sendNoContextTakeover: Bool
+        let compressionLevel: Int?
+        let memoryLevel: Int?
+    }
+
+    /// Internal mutable state and referenced types, that cannot be set to Sendable
+    class InternalState {
+        fileprivate let decompressor: any NIODecompressor
+        fileprivate let compressor: any NIOCompressor
+        fileprivate var sendState: SendState
+
+        init(configuration: Configuration) throws {
+            self.decompressor = CompressionAlgorithm.deflate(
+                configuration: .init(
+                    windowSize: numericCast(configuration.receiveMaxWindow ?? 15)
+                )
+            ).decompressor
+            // compression level -1 will setup the default compression level, 8 is the default memory level
+            self.compressor = CompressionAlgorithm.deflate(
+                configuration: .init(
+                    windowSize: numericCast(configuration.sendMaxWindow ?? 15),
+                    compressionLevel: configuration.compressionLevel.map { numericCast($0) } ?? -1,
+                    memoryLevel: configuration.memoryLevel.map { numericCast($0) } ?? 8
+                )
+            ).compressor
+            self.sendState = .idle
+            try self.decompressor.startStream()
+            try self.compressor.startStream()
+        }
+
+        func shutdown() {
+            try? self.compressor.finishStream()
+            try? self.decompressor.finishStream()
+        }
+    }
+
+    let configuration: Configuration
+    let internalState: NIOLoopBound<InternalState>
+
+    init(configuration: Configuration, eventLoop: EventLoop) throws {
+        self.configuration = configuration
+        self.internalState = try .init(.init(configuration: configuration), eventLoop: eventLoop)
+    }
+
+    func shutdown() {
+        self.internalState.value.shutdown()
+    }
+
+    func processReceivedFrame(_ frame: WebSocketFrame, ws: HBWebSocket) throws -> WebSocketFrame {
+        var frame = frame
+        if frame.rsv1 {
+            let state = self.internalState.value
+            precondition(frame.fin, "Only concatenated frames with fin set can be processed by the permessage-deflate extension")
+            frame.data.writeBytes([0, 0, 255, 255])
```

Comment:
What are these constants?

Reply:
I should have added a link to the specific lines of the RFC here. It requires an empty deflate block: `00,00,ff,ff` to be added to the end of the compressed data before decompression. 

There is an equivalent operation in the compression where this block is removed.

I'll add a comment